README.md
# crc-flex

This library designed to allow the calculation of CRCs for arbitrary configurations of width, underlying polynomial, initial value, final XOR and reflecting of the input / output.

This module is based on the information and subsequent C code implementation
given ["A PAINLESS GUIDE TO CRC ERROR DETECTION ALGORITHM"](http://read.pudn.com/downloads137/doc/585979/crc-Ross.pdf), Ross N. Williams, 19th August 1993.

The library is focused on the flexibility to model many CRCs easily and is not optimised for speed or memory usage.


ChangeLog.md
# Changelog for crc-model
0.1.0.0
* Initial release

## Unreleased changes


test/Spec.hs

import Test.Hspec
import qualified Data.ByteString.Char8 as BSC
import qualified Data.ByteString as BS
import Data.Digest.CrcFlex
import Data.Digest.CrcFlex.Internal
import Data.Word

testinput :: [Word8]
testinput = BS.unpack $ BSC.pack "123456789"

builtinChecks :: [(String, CrcModel, Integer)]
builtinChecks = [
          ("crc-7 (MMC)", crc7Mmc, 0x75)
        , ("crc-8 (MAXIM)", crc8Maxim, 0xa1)
        , ("crc-11 (FLEXRAY)", crc11FlexRay, 0x5a3)
        , ("crc-16 (CCITT)", crcCCITT, 0x29b1)
        , ("crc-16 (IBM)", crc16IBM, 0xbb3d)
        , ("crc-32 (ISO)", crc32Iso, 0xcbf43926)
        , ("crc-64 (ISO)", crc64Iso, 0xb90956c775a41001)
        , ("crc-64 (EMCA)", crc64Emca, 0x6c40df5f0b497347)
    ]

main :: IO ()
main = hspec $ do
    describe "Data.Digest.CrcFlex.Internal.pow2" $ do
        it "produces 2^n for positive numbers" $ 
            pow2 8 `shouldBe` (256 :: Int)
        it "should work for n == 0" $
            pow2 0 `shouldBe` (1 :: Int)

    describe "Data.Digest.CrcFlex.Internal.reflect" $ do
        it "can reflect an 8-bit integer" $ 
            reflect 8 0xA3 `shouldBe` (0xC5 :: Integer)
        it "can reflect a 16-bit integer" $ 
            reflect 16 0x1A23 `shouldBe` (0xC458 :: Integer)
        it "reflects the least significant bits" $ 
            reflect 8 0x1A23 `shouldBe` (0x1AC4 :: Integer)

    describe "Data.Digest.CrcFlex.applyCrcModel" $ 
            mapM_ (\(n, c,chk) -> it ("should work for " ++ n) (applyCrcModel c testinput `shouldBe` chk)) builtinChecks
   
    describe "Data.Digest.CrcFlex.applyCrcTable" $
        mapM_ (\(n, c,chk) -> it ("should work for " ++ n) (applyCrcTable (mkCrcTable c) testinput `shouldBe` chk)) builtinChecks
        

src/Data/Digest/CrcFlex/Internal.hs
module Data.Digest.CrcFlex.Internal where

import Data.Array (Array, listArray, (!))
import Data.Bits ((.&.), xor, shift, shiftR, testBit, complement)
import Data.List (foldl')

-- |The 'CrcModel' holds all of the information to model a CRC algorithm.
data CrcModel = CrcModel {
    cmWidth :: Int,      -- ^ specifies the width (in bits) of the CRC
    cmPoly :: Integer,   -- ^ specifies the polynomial to be used for the CRC calculation
    cmInit :: Integer,   -- ^ specifies the initial vlaue for the remainder for the CRC calculation
    cmReflect :: Bool,   -- ^ specifies whether the CRC uses the "reflected" form
    cmXor :: Integer     -- ^ specifies a value to XOR the final remainder (after reflection) with during the CRC calculation
}

-- |The 'CrcTable' holds the information required to generate a CRC using a table driven approach
data CrcTable = CrcTable {
    ctModel :: CrcModel,
    ctNext :: Integer -> Integer -> Integer
} 

-- |The 'pow2' function produces the 2^n for a given n.
-- This only works for n >= 0.
pow2 :: (Num b, Enum b, Eq b) => Int -> b
pow2 = (map pow2' [0..] !!)
    where
        pow2' :: (Num b, Enum b, Eq b) => Int -> b
        pow2' 0 = 1 
        pow2' n = 2 * pow2' (n-1)

-- |The 'reflect' function reverses the n least significant bits of x.
reflect :: Int -> Integer -> Integer
reflect n x = (x .&. mask) + refbits
    where
        ms = pow2 <$> reverse [0..(n-1)]
        ss = take n $ iterate (+2) (-(n-1))
        mask = complement $ pow2 n - 1
        refbits = sum $ zipWith (\m s -> shift (x .&. m) s) ms ss

-- |The 'toBits' function returns the bit pattern for the input (x)
-- It needs to be passed the input width (bits, w)
-- The resulting bit stream is represented as [Bool] and is ordered lsb to msb
toBits :: Int -> Integer -> [Bool]
toBits w x = [testBit x i | i <- [0..w-1]]

-- |The 'nextdigestbit' function uses CrcModel (m), the current remainder (x) and the next bit (b) to produces the next remainder based on the CRC polynomial.
nextdigestbit :: CrcModel -> Integer -> Bool -> Integer
nextdigestbit m x b = if (/=) (testBit x (cmWidth m - 1)) b then shift x 1 `xor` cmPoly m else shift x 1

-- |The 'nextdigest' function takes CrcModel m, the current CRC x and the next byte x
-- It iterates (using the CrcModel) for each of the 8 bits of the input byte and returns the resulting remainder.
nextdigest :: CrcModel -> Integer -> Integer -> Integer
nextdigest m crc x = crc' .&. wndmask
    where
        wndmask = pow2 (cmWidth m) - 1
        x' = if not (cmReflect m) then fromInteger (reflect 8 x) else fromInteger x
        crc' = foldl' (nextdigestbit m) crc $ toBits 8 x'

-- |Used to update the remainder (for normal form CrcModels - i.e. cmReflect == False) based on the next byte and a lookup table.
nextTableByte :: Array Int Integer -> Int -> Integer -> Integer -> Integer
nextTableByte ts w crc x = (shift crc 8 `xor` (ts ! index)) .&. wndmask
    where
        wndmask = pow2 w - 1
        highbyte = shiftR crc (w - 8)
        index = fromInteger (x `xor` highbyte)

-- |Used to update the remainder (for reversed form CrcModels - i.e. cmReflect == True) based on the next byte and a lookup table.
nextTableByteReflected :: Array Int Integer -> Int -> Integer -> Integer -> Integer
nextTableByteReflected ts w crc x = (shiftR crc 8 `xor` (ts ! index)) .&. wndmask
    where
        wndmask = pow2 w - 1
        lowbyte = crc .&. 0xFF       
        index = fromInteger (x `xor` lowbyte) 

-- |Generates a CRC table for a given CrcModel.
mkCrcTable :: CrcModel -> CrcTable
mkCrcTable m = CrcTable m nxt
    where
        genIndex = nextdigest m 0
        mayReflect = if cmReflect m then reflect (cmWidth m) else id
        ts = listArray (0, 255) $ map (mayReflect . genIndex) [0..255]
        nxt = if cmReflect m then nextTableByteReflected ts (cmWidth m) else nextTableByte ts (cmWidth m)

src/Data/Digest/CrcFlex.hs
{-|
Module      : CrcFlex
Description : Computes arbitrary CRCs over a stream of [Word8]
Copyright   : (c) Paul Randall, 2020
License     : MIT
Maintainer  : paul-randall2@sky.com
Stability   : experimental
Portability : POSIX

This module is based on the information and subsequent C code implementation
given "A PAINLESS GUIDE TO CRC ERROR DETECTION ALGORITHM", Ross N. Williams, 19th August 1993.
http://read.pudn.com/downloads137/doc/585979/crc-Ross.pdf

It is designed to allow the calculation of CRCs for arbitrary CRC widths where the underlying 
polynomial and options for initial value, final XOR and reflecting of the input / output can be
independently controlled.

It is not designed to be efficient in terms of speed or memory usag more to provide enough flexibility to 
cater to the lesser used CRCs that are not implemented in other packages.

There is a good catlog of CRCs and their implementation details here:-
https://reveng.sourceforge.io/crc-catalogue/
-}

module Data.Digest.CrcFlex (CrcModel(..), CrcTable(..), applyCrcModel, mkCrcTable, applyCrcTable
        , crc7Mmc, crc11FlexRay, crc8Maxim, crc16IBM, crcCCITT, crc32Iso, crc64Iso, crc64Emca
        ) where

import Data.Digest.CrcFlex.Internal
import Data.Bits (xor)
import Data.List (foldl')
import Data.Word (Word8)

-- |The 'crc7Mmc' function provides the model for the crc-7 (used for MMC devices).
crc7Mmc :: CrcModel
crc7Mmc = CrcModel 7 0x09 0 False 0

-- |The 'crc8Maxim' function provides the model for the Maxim crc-8 (used for 1-wire devices).
crc8Maxim :: CrcModel
crc8Maxim = CrcModel 8 0x31 0 True 0

-- |The 'crc11FlexRay' function provides the model for the FlexRay crc-11.
crc11FlexRay :: CrcModel
crc11FlexRay = CrcModel 11 0x385 0x01a False 0 

-- |The 'crc16IBM' function provides the model for the commonly used IBM crc-16.
crc16IBM :: CrcModel
crc16IBM = CrcModel 16 0x8005 0 True 0

-- |The 'crcCCITT' function provides the model for the commonly used CRC-16-CCITT.
crcCCITT :: CrcModel
crcCCITT = CrcModel 16 0x1021 0xFFFF False 0

-- |The 'crc32Iso' function provides the model for the commonly used CRC-32 (ISO).
crc32Iso :: CrcModel
crc32Iso = CrcModel 32 0x04C11DB7 0xFFFFFFFF True 0xFFFFFFFF

-- |The 'crc64Iso' function provides the model for the commonly used CRC-64 (ISO).
crc64Iso :: CrcModel
crc64Iso = CrcModel 64 0x1B 0xffffffffffffffff True 0xffffffffffffffff

-- |The 'crc64Emca' function provides the model for the commonly used CRC-64 (ISO).
crc64Emca :: CrcModel
crc64Emca = CrcModel 64 0x42f0e1eba9ea3693 0 False 0

-- |The 'applyCrcModel' function produces the resulting CRC from running the CrcModel over the input [Word8] stream.
applyCrcModel :: CrcModel -> [Word8] -> Integer
applyCrcModel m xs = c' `xor` cmXor m
    where
        c = foldl' (nextdigest m) (cmInit m) $ map fromIntegral xs
        c' = if cmReflect m then reflect (cmWidth m) c else c

-- |The 'applyCrcTable' function produces the resulting CRC by applying the CrcTable over the input [Word8] stream.
applyCrcTable :: CrcTable -> [Word8] -> Integer
applyCrcTable (CrcTable m nxt) xs = c `xor` cmXor m
    where
        st = if cmReflect m then reflect (cmWidth m) (cmInit m) else cmInit m
        c = foldl' nxt st $ map fromIntegral xs